src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,193 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.FindSymbols;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.InlineMethod
+{
+    internal abstract partial class AbstractInlineMethodRefactoringProvider : CodeRefactoringProvider
+    {
+        private readonly ISyntaxFacts _syntaxFacts;
+
+        protected abstract Task<SyntaxNode?> GetInvocationExpressionSyntaxNodeAsync(CodeRefactoringContext context);
+
+        /// <summary>
+        /// Check if the <param name="calleeMethodDeclarationSyntaxNode"/> has only one expression or it is using arrow expression.
+        /// </summary>
+        protected abstract bool IsMethodContainsOneStatement(SyntaxNode calleeMethodDeclarationSyntaxNode);
+
+        protected abstract SyntaxNode? GetInlineStatement(SyntaxNode calleeMethodDeclarationSyntaxNode);
+
+        protected AbstractInlineMethodRefactoringProvider(ISyntaxFacts syntaxFacts)
+        {
+            _syntaxFacts = syntaxFacts;
+        }
+
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var (document, _, cancellationToken) = context;
+            var calleeMethodInvocationSyntaxNode = await GetInvocationExpressionSyntaxNodeAsync(context).ConfigureAwait(false);
+            if (calleeMethodInvocationSyntaxNode == null)
+            {
+                return;
+            }
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+            if (semanticModel == null)
+            {
+                return;
+            }
+
+            var methodSymbol = TryGetBestMatchSymbol(semanticModel, calleeMethodInvocationSyntaxNode, cancellationToken);
+            if (methodSymbol == null
+                || methodSymbol.DeclaredAccessibility != Accessibility.Private
+                || !methodSymbol.IsOrdinaryMethod())
+            {
+                return;
+            }
+
+            if (methodSymbol is IMethodSymbol calleeMethodInvocationSymbol)
+            {
+                var calleeMethodDeclarationSyntaxNodes = await Task.WhenAll(calleeMethodInvocationSymbol.DeclaringSyntaxReferences
+                    .Select(reference => reference.GetSyntaxAsync())).ConfigureAwait(false);
```

Comment:
1. no need for whenall here.
2. pass cancellation token to GetSyntaxAsync
3. you can check the length before doing this.